conn pool: use hostnames of endpoints as SNI values

dmitriyilin opened the pull request:
Commit Message: conn pool: use hostnames of endpoints as SNI values
Additional Description: optional support for usage of upstream cluster endpoints' hostnames as SNI values. This PR is the successor of [34898](https://github.com/envoyproxy/envoy/pull/34898) with expected improvement of test flakiness. Original PR was reverted by [35212](https://github.com/envoyproxy/envoy/pull/35212).
Risk Level: Low
Testing: integration
Docs Changes: added information about new mechanism of SNI derivation
Release Notes: https://github.com/dmitriyilin/envoy/blob/0f70012a2383dd56cb070199664be44dbd8bbd93/changelogs/current.yaml#L16
Platform Specific Features: N/A
Fixes https://github.com/envoyproxy/envoy/issues/15839

dmitriyilin commented:
fyi, I took "ProxyFilterIntegrationTest, UpstreamTlsInvalidSAN" as basis for the test that was showing flaky behavior "AutoSniIntegrationTest, AutoSniFromUpstreamAndAutoSanValidationFailureTest".

dmitriyilin commented:
@adisuissa , could you "rereview" this change?

adisuissa reviewed (commented):
/lgtm api

adisuissa commented:
Assigning Matt who also reviewed #34898. Can you provide more info about how the flaky test was fixed?
/assign @mattklein123

yanavlasov commented on source/common/conn_pool/conn_pool_base.cc:
```diff
@@ -19,13 +19,44 @@ namespace {
+      !upstream_http_protocol_options->auto_sni_from_upstream()) {
+    return transport_socket_options;
+  }
+  const absl::string_view hostname = host->hostname();
```
I believe `host->hostname()` can include `:port` if it was resolved by the Dynamic Forward Proxy. You can end up with invalid SNI.

dmitriyilin replied:
According to the comments `hostname` is a hostname and there is also `address` available in the `HostDescription`, so it would be very confusing to also include a port in `hostname`. I also can not spot where a port can be added to a hostname of a `HostImpl` or a `LogicalHost`.

yanavlasov reviewed (commented):
/wait

dmitriyilin commented:
> Assigning Matt who also reviewed #34898. Can you provide more info about how the flaky test was fixed? /assign @mattklein123

I had used `ProxyFilterIntegrationTest. UpstreamTlsInvalidSAN` test as the basis and added missed configuration:
[FakeUpstream.setReadDisableOnNewConnection(false)](https://github.com/dmitriyilin/envoy/blob/055a280f2e25c72057027d137ca9e0930f81a9d4/test/extensions/filters/http/router/auto_sni_integration_test.cc#L155)

As flaky test validates behaviour in case of failure, I've added explicit cleanup:
[HttpIntegrationTest.cleanupUpstreamAndDownstream()](https://github.com/dmitriyilin/envoy/blob/055a280f2e25c72057027d137ca9e0930f81a9d4/test/extensions/filters/http/router/auto_sni_integration_test.cc#L164)

dmitriyilin commented:
@yanavlasov, @adisuissa, @mattklein123, just a reminder

quantumsheep commented:
Hey! This is exactly the feature I'm needing. Do you think it's possible to do the same using current Envoy's release configuration?

dmitriyilin commented:
> Hey! This is exactly the feature I'm needing. Do you think it's possible to do the same using current Envoy's release configuration?

It's a new functionality, so new config is needed.

dmitriyilin commented:
@yanavlasov, @adisuissa, @mattklein123 do you see any issues with the changes?

dmitriyilin commented:
@mattklein123 @yanavlasov @adisuissa any feedback?

mattklein123 commented:
Can you please merge main and I can take a look?

/wait

ggreenway reviewed (commented):
I didn't see this PR and didn't know it existed. I implemented roughly the same thing in #36903, which merged last week.

dmitriyilin commented:
@ggreenway , thanks for the info. I am closing this PR then
